### PR TITLE
 Ensuring proper resource reconciliation on `Repository` Creation

### PR DESCRIPTION
**Pull request by a-hilaly**
Resolves: https://github.com/aws-controllers-k8s/community/issues/1873

**TLDR:**
- Addressed a bug in the ECR controller where Repository Custom Resources (CRs) failed
  to create `Policy` and `LifecyclePolicy` entries when containing all (possible) fields.
- The ECR controller was missing the post `sdkCreate` hooks; But the resources that 
  were missing few default-able fields we properly reconciled primarily due to late
  initialization, detecting API default values for non-specified fields, causing the controller
  to re queue resources after they have been created.
- Resolved by adding the necessary hooks to the create path.

**Description:**
This fix addresses a bug in the ECR controller. When a user submitted a Repository
Custom Resource (CR) containing all fields, the controller failed to create `Policy`
and `LifecyclePolicy` entries. However, if one of the fields was missing (excluding
`Policy` or `LifecyclePolicy`), the reconciliation process succeeded.

The bug fix was straightforward and intuitive, as it became evident that the controller
lacked the required hook for the create path. However, this raised questions about why
`policy` and `LifecyclePolicy` were missing in some specific cases. Our e2e tests
created CRs with isolated fields, such as `Policy` and `LifecyclePolicy`, and verified
that these policies were indeed applied to the ECR Repository. 

It was evident that the reason these specific cases were successfully reconciled was
due to the correct logic in the update path. This logic ensured that if a `Policy` or
`LifecyclePolicy` was missing, an API call was made to update these fields.

The mystery revolved around why these edge cases were reconciled even when we didn't
explicitly request a second reconciliation or requeue of the resource

**Debugging:**

- Our debugging journey began with a review of the logs and a comparison between the logs
  of a resource created with all fields and one with only the policy. Surprisingly, the
  logs showed a complete aditional reconciliation step, and it wasn't initiated by the controller
  itself (no requeue, no drift remediation, no condition). This was puzzling.

- To gain deeper insights, we connected the controller to a local clone of the runtime and
  added print statements. However, this approach made the logs messier and harder to
  interpret. Consequently, we tried to debugging with a go-delve debugger.

- Our debugging session took us through the ACK runtime, but the issue didn't originate
  there. Instead, it originiated from 'k8s-sigs/controller-runtime.' We cloned it and extended
  our debugging to the `controller-runtime`. There, we discovered that 'c.Do.Reconcile(ctx, req)'
  was causing two reconciliations.

- Further investigation revealed that 'for c.processNextWorkItem(ctx) {}' was called,
  indicating the presence of an extra item in the queue. This item had to be an update
  because the first item in the queue was a "create".

- The delve debugger showed that we submitted two resources one with only few fields
  and the other with all the fields... We then ran a `diff` on `k get repository <repo1/repo2> -o yaml`
  which revealed a crucial clue: 'metadata.generation' was 1 for the policy+name-only
  resource and 2 for the all the fields resource, confirming the update theory.

- The controller initiated the resource update because the ECR API returned all fields,
  even those not explicitly provided. This behavior prompted the ACK runtime to execute
  a patch into the resource spec, incrementing the generation and triggering an update.

Yeah, we know what that late initialisation is a thing, but it completly baited me here...

Signed-off-by: Amine Hilaly <hilalyamine@gmail.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/e2e/tests/test_repository.py [hunk 1]
```diff
@@ -365,3 +366,56 @@ def test_repository_policy(self, ecr_client):
...
+        k8s.create_custom_resource(ref, resource_data)
+        cr = k8s.wait_resource_consumed_by_controller(ref)
+
+        assert cr is not None
+        assert k8s.get_resource_exists(ref)
+
+        time.sleep(CREATE_WAIT_AFTER_SECONDS)
+
+        # Get latest repository CR
+        cr = k8s.wait_resource_consumed_by_controller(ref)
```
Comment (a-hilaly) [inline test/e2e/tests/test_repository.py:400]: I'm aware that this ideally should be a fixture, but we probably want to do in a separate PR to refactor all the e2e tests in the ecr-controller and other controllers as well..
